File: Leetcode/BFS/#116. Populating Next Right Pointers in Each Node.cpp
//DFS
class Solution {
public:
    void help(Node* root, Node* n, int LR){
        if(!root) return;
        
        if(LR==1) root->next=n->right;
        else if(LR==2){
           if(n->next) root->next=n->next->left;
        }
        
        help(root->left,root,1);
        help(root->right,root,2);
    }
    
    Node* connect(Node* root) {
        help(root,NULL,3);
        return root;
    }
};

//BFS
class Solution {
public:
    
    Node* connect(Node* root) {
    if(!root) return root;
       queue<Node*>Q;
        Q.push(root);
        
        while(!Q.empty()){
            int n = Q.size();
            int cnt=0;
            for(int i=0; i<n; i++){
                Node* cur = Q.front(); Q.pop(); 
                if(i!=n-1) cur->next=Q.front(); cnt++;
                if(cur->left) Q.push(cur->left);
                if(cur->right) Q.push(cur->right);
            }
        }
        
        return root;
    }
};

File: BFS/#116. Populating Next Right Pointers in Each Node.py

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None
        res = root
        Q = []
        Q.append(root)

        while Q:
            prev = None
            level_size = len(Q)
            for _ in range(level_size):
                curr = Q[0]
                Q.pop(0)
                if not prev: # leftmost
                    prev = curr
                else: # in-between
                    prev.next = curr
                    prev = curr
                
                if curr.left:
                    Q.append(curr.left)
                if curr.right:
                    Q.append(curr.right)
        return res